src/passes/LoopInvariantCodeMotion.cpp

```diff
@@ -0,0 +1,246 @@
+    // Walk along the loop entrance, while all the code there
+    // is executed unconditionally. That is the code we want to
+    // move out - anything that might or might not be executed
+    // may be best left alone anyhow.
+    std::vector<Expression**> work;
+    work.push_back(&loop->body);
+    while (!work.empty()) {
+      auto** currp = work.back();
+      work.pop_back();
+      auto* curr = *currp;
+      // Look into blocks.
+      if (auto* block = curr->dynCast<Block>()) {
+        auto& list = block->list;
+        Index i = list.size();
+        if (i > 0) {
+          do {
+            i--;
+            work.push_back(&list[i]);
+          } while (i != 0);
+        }
+        continue;
+        // Note that if the block had a merge at the end, we would have seen
+        // a branch to it anyhow, so we would stop before that point anyhow.
+      }
+      // If this may branch, we are done.
+      EffectAnalyzer effects(getPassOptions(), curr);
+      if (effects.branches) {
+        break;
+      }
+      if (interestingToMove(curr)) {
+        // Let's see if we can move this out.
+        // Global side effects would prevent this - we might end up
+        // executing them just once.
+        // And we must also move across anything not moved out already,
+        // so check for issues there too.
+        // The rest of the loop's effects matter too, we must also
+        // take into account global state like interacting loads and
+        // stores.
+        if (!effects.hasGlobalSideEffects() &&
+            !effectsSoFar.invalidates(effects) &&
+            !(effects.noticesGlobalSideEffects() && loopEffects.hasGlobalSideEffects())) {
+          // So far so good. Check if our local dependencies are all
+          // outside of the loop, in which case everything is good -
+          // either they are before the loop and constant for us, or
+          // they are after and don't matter.
+          bool canMove = true;
+          if (!effects.localsRead.empty()) {
+            FindAll<GetLocal> gets(curr);
+            for (auto* get : gets.list) {
+              auto& sets = localGraph->getSetses[get];
+              for (auto* set : sets) {
+                // nullptr means a parameter or zero-init value;
+                // no danger to us.
+                if (!set) continue;
+                // Check if the set is in the loop. If not, it's either before,
+                // which is fine, or after, which is also fine - moving curr
+                // to just outside the loop will preserve those relationships.
+                // TODO: this still counts curr's sets as inside the loop, which
+                //       might matter in non-flat mode.
+                if (loopSets.count(set)) {
+                  canMove = false;
+                  break;
+                }
+              }
+              if (!canMove) {
+                break;
+              }
+            }
+          }
+          if (canMove) {
```

Comment:
All these `if (canMove)`s are really awkward control flow. This loop is also getting really deeply indented. Can probably fix both of these by extracting some functions here, which can turn this loop into something like:
```c++
if (interestingToMove(curr)) {
  if (!effects.hasGlobalSideEffects() &&
      !effectsSoFar.invalidates(effects) &&
      !(effects.noticesGlobalSideEffects() && loopEffects.hasGlobalSideEffects())) {
    bool canMove = getsAreOk(curr) && setsAreOk(curr, numSetsForIndex);
    if (canMove) {
      // We can move it! ...
    }
  }
}
```

This also lets us say things like `return false` instead of `canMove = false; break` in those functions, which should simplify things more.

Reply:
Good point, fixing.